quota/quota.go

```diff
@@ -0,0 +1,315 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package quota
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+)
+
+// Group is a quota group of snaps, services or sub-groups that are all subject
+// to specific resource quotas. The only quota resource types currently
+// supported is memory, but this can be expanded in the future.
+type Group struct {
+	// Name is the name of the quota group. Certain names are reserved and have
+	// special meaning such as user.* or system.* or snapd.*, but are otherwise
+	// generically choosable by a user. This name will correspond to the name of
+	// the systemd slice underlying the quota group.
+	Name string `json:"name,omitempty"`
+
+	// SubGroups is the set of sub-groups that are subject to this quota.
+	// Sub-groups have their own limits, subject to the requirement that the
+	// highest quota for a sub-group is that of the parent group.
+	SubGroups []string `json:"sub-groups,omitempty"`
+
+	// subGroups is the set of actual sub-group objects, needed for tracking and
+	// calculations
+	subGroups []*Group
+
+	// MemoryLimit is the limit of memory available to the processes in the
+	// group where if the total used memory of all the processes exceeds the
+	// limit, oom-killer is invoked which will start killing processes. The
+	// specific behavior of which processes are killed is subject to the
+	// ExhaustionBehavior.
+	MemoryLimit quantity.Size `json:"memory-limit,omitempty"`
+
+	// MemoryExhaustionBehavior is the behavior that is implemented to decide
+	// what processes to kill in the quota group when the memory limits are
+	// exhausted.
+	MemoryExhaustionBehavior MemoryExhaustionBehavior `json:"memory-exhaustion-behavior,omitempty"`
```

Comment:
Maybe just OOMBehavior / oom-behavior?

Reply:
@pedronis thoughts on this one?

Reply:
it's a bit unclear whether we should introduce this at all given it has only one possible value atm,  it's JSON is it would be easy to introduce it later when we have more than one value, and use no value as what the initial behavior was.